File: cpp/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> stk;

        for (auto ast: asteroids)
        {
            while (!stk.empty() && stk.back() > 0 && ast < 0)
            {
                int diff = ast + stk.back();
                if (diff > 0)
                {
                    ast = 0;
                }
                else if (diff < 0)
                {
                    stk.pop_back();
                }
                else
                {
                    ast = 0;
                    stk.pop_back();
                } 
            }
            if (ast != 0)
                stk.push_back(ast);
        }
        return stk;
    }
};

File: kotlin/0735-asteroid-collision.kt
class Solution {
    fun asteroidCollision(ast: IntArray): IntArray {
        val stack = LinkedList<Int>()

        for (_a in ast) {
            var a = _a
            while (stack.isNotEmpty() && a < 0 && stack.peekLast() > 0) {
                val diff = a + stack.peekLast()
                if (diff < 0) {
                    stack.removeLast()
                } else if (diff > 0) {
                    a = 0
                } else {
                    a = 0
                    stack.removeLast()
                }
            }

            if (a != 0) stack.addLast(a)
        }

        return stack.toIntArray()
    }
}
